graylog2-server/src/main/java/org/graylog/events/CheckEventDefinitionPermissions.java

```diff
@@ -0,0 +1,115 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.events;
+
+import org.graylog.events.processor.EventDefinitionDto;
+import org.graylog.grn.GRNRegistry;
+import org.graylog.grn.GRNTypes;
+import org.graylog.plugins.views.search.errors.PermissionException;
+import org.graylog.security.Capability;
+import org.graylog.security.DBGrantService;
+import org.graylog.security.UserContext;
+
+import static org.graylog2.shared.security.RestPermissions.EVENT_DEFINITIONS_CREATE;
+import static org.graylog2.shared.security.RestPermissions.EVENT_DEFINITIONS_DELETE;
+import static org.graylog2.shared.security.RestPermissions.EVENT_DEFINITIONS_EDIT;
+import static org.graylog2.shared.security.RestPermissions.EVENT_DEFINITIONS_EXECUTE;
+import static org.graylog2.shared.security.RestPermissions.EVENT_DEFINITIONS_READ;
+
+public class CheckEventDefinitionPermissions {
+    private final UserContext userContext;
+    private final DBGrantService dbGrantService;
+    private final GRNRegistry grnRegistry;
+
+    private CheckEventDefinitionPermissions(final UserContext userContext, final DBGrantService dbGrantService, final GRNRegistry grnRegistry) {
+        this.userContext = userContext;
+        this.dbGrantService = dbGrantService;
+        this.grnRegistry = grnRegistry;
+    }
+
+    public static CheckEventDefinitionPermissions create(final UserContext userContext, final DBGrantService dbGrantService, final GRNRegistry grnRegistry) {
+        return new CheckEventDefinitionPermissions(userContext, dbGrantService, grnRegistry);
+    }
+
+    // check for nonEmpty, as the minimal grant is view capability
+    private boolean readAllowedOnEntity(final String id) {
+        final var targetGRN = grnRegistry.newGRN(GRNTypes.EVENT_DEFINITION, id);
+        final var userGRN = grnRegistry.newGRN(GRNTypes.USER, userContext.getUser().getId());
+        return !dbGrantService.getForTargetAndGrantee(targetGRN, userGRN).isEmpty();
+    }
+
+    private boolean manageAllowedOnEntity(final String id) {
+        final var targetGRN = grnRegistry.newGRN(GRNTypes.EVENT_DEFINITION, id);
+        final var userGRN = grnRegistry.newGRN(GRNTypes.USER, userContext.getUser().getId());
+        final var grants = dbGrantService.getForTargetAndGrantee(targetGRN, userGRN);
+        return grants.stream().anyMatch(g -> g.capability().priority() >= Capability.MANAGE.priority());
+    }
+
+    public void checkRead(final String id) {
+        if(!canRead(id)) {
+            throw new PermissionException("Action not allowed");
+        }
+    }
+
+    public boolean canRead(final String id) {
+        return userContext.isPermitted(EVENT_DEFINITIONS_READ, id) || readAllowedOnEntity(id);
```

Comment:
Why do we need the `readAllowedOnEntity` method? If the user in `userContext` has a grant to view an entity, it will be reflected in their permissions and `isPermitted()` will work. Do I miss anything?

Reply:
thx. for pointing that out. I will review/change the code. The premise was: I need two checks, one for the permission on a role (e.g. manager role) and another check for the entity/sharing mechanism. It's probable that I implemented it wrong.